src/core/net/tcp6_ext.cpp

```diff
@@ -0,0 +1,234 @@
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements TCP/IPv6 socket extensions.
+ */
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_TCP_ENABLE
+
+#include "tcp6_ext.hpp"
+
+#include "common/as_core_type.hpp"
+#include "common/code_utils.hpp"
+#include "common/error.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+#include "common/log.hpp"
+#include "common/random.hpp"
+#include "net/checksum.hpp"
+#include "net/ip6.hpp"
+#include "net/netif.hpp"
+
+#include "../../third_party/tcplp/tcplp.h"
+
+namespace ot {
+namespace Ip6 {
+
+RegisterLogModule("TcpExt");
+
+void TcpCircularSendBuffer::Initialize(void *aDataBuffer, size_t aCapacity)
+{
+    this->mDataBuffer = static_cast<uint8_t *>(aDataBuffer);
+    this->mCapacity   = aCapacity;
+    ForceDiscardAll();
+}
+
+Error TcpCircularSendBuffer::Write(Tcp::Endpoint &aEndpoint,
+                                   const void *   aData,
+                                   size_t         aLength,
+                                   size_t &       aWritten,
+                                   uint32_t       aFlags)
+{
+    Error error = kErrorNone;
+
+    size_t bytesFree = FreeSpace();
+    size_t writeIndex;
+
+    uint32_t flags = 0;
+    size_t   bytesUntilWrap;
+
+    VerifyOrExit(aLength != 0);
+    VerifyOrExit(bytesFree != 0, aLength = 0);
+
+    /*
+     * Handle the case where we don't have enough space to accommodate all of the
+     * provided data.
+     */
+    if (aLength > bytesFree)
+    {
+        aLength = bytesFree;
+    }
+
+    /*
+     * This is a "simplifying" if statement the removes an edge case from the logic
+     * below. It guarantees that a write to an empty buffer will never wrap.
+     */
+    if (this->mCapacityUsed == 0)
+    {
+        this->mStartIndex = 0;
+    }
+
+    writeIndex = GetIndex(this->mStartIndex, this->mCapacityUsed);
+
+    if ((aFlags & OT_TCP_CIRCULAR_SEND_BUFFER_WRITE_MORE_TO_COME) != 0)
+    {
+        flags |= OT_TCP_SEND_MORE_TO_COME;
+    }
+
+    bytesUntilWrap = GetNumBytesUntilWrap(writeIndex);
+    if (aLength <= bytesUntilWrap)
+    {
+        memcpy(&this->mDataBuffer[writeIndex], aData, aLength);
+        if (this->mCapacityUsed == 0 || writeIndex == 0)
```

Comment:
Can we remove the `mCapacityUsed == 0` check here?

If I am not mistaken, when `mCapacityUsed` is zero from the `if()` above we do `mStartIndex = 0` which should then cause `writeIndex = GetIndex(0, 0)` to also be zero.

Reply:
That sounds correct to me.